task2/task2.1.js

```diff
@@ -0,0 +1,171 @@
+const { v4: get_uuid} = require('uuid');
+const express = require('express');
+
+const app = express();
+app.use(express.json());
+
+app.use(function(req, _res, next) {
+    console.log(req.body);
+    next();
+});
+
+class User {
+    constructor(id, login, password, age, isDeleted = false) {
+        this.id = id ?? get_uuid();
+        this.login = login;
+        this.password = password;
+        this.age = age;
+        this.isDeleted = isDeleted;
+    }
+}
+
+class Users {
+    constructor() {
+        this.users = [];
+    }
+
+    get(id) {
+        return id && this.users.find(u => u.id === id);
+    }
+
+    getNotDeleted(id) {
+        const user = this.get(id);
+        return user && user.isDeleted !== true ? user : undefined;
+    }
+
+    create({id, login, password, age}) {
+        if (this.get(id)) {
+            throw new Error(`Cannot create - user id ${id} already exists`);
+        }
+
+        if (this.users.some(u => u.login === login)) {
+            throw new Error(`Cannot create - user with the same login ${login} already exists`);
+        }
+        const newUser = new User(id, login, password, age);
+        this.users.push(newUser);
+        return newUser.id;
+    }
+
+    update({id, login, password, age}) {
+        const user = this.getNotDeleted(id);
+        if (!user) {
+            throw new Error(`Cannot update - user id ${id} does not exists or removed`);
+        }
+
+        if (login) user.login = login;
+        if (password) user.password = password;
+        if (age) user.age = age;
+    }
+
+    delete(id, soft = true) {
+        const user = this.get(id);
+        if (!user) {
+            throw new Error(`Cannot remove - user id ${id} does not exists`);
+        }
+        if (soft) {
+            user.isDeleted = true;
+        } else {
+            this.users = this.users.filter(u => u.id !== id);
+        }
+    }
+
+    getAutoSuggestUsers(loginSubstring, limit = 0) {
+        return this.users
+            .filter(u => !u.isDeleted && u.login && u.login.includes(loginSubstring))
+            .sort((u1, u2) => u1.login > u2.login ? 1 :
+                            u1.login < u2.login ? -1 : 0)
+            .slice(0, limit);
+    }
+
+    getAllUsers() {
+        return this.users;
+    }
+}
+
+const users_db = new Users();
+
+app.get('/api/user', function(_req, res) {
+    res.json({users: users_db.getAllUsers()});
+});
+
+function verifyQueries(params, req, res, next) {

```

Comment:
you can convert it to HOF
```javascript
function verifyQueries(params) {
  return function(req, res, next) {
    ...
  }
}

```

also, it's better to use error-first approach to avoid unnecessary nesting
and one more thing: `req` and `req.quey` will always be presented since it's middleware